### PR TITLE
build: publish benchmark-only classifier jars (blob, sts)

### PR DESCRIPTION
**Pull request by iamabhilaksh**
## What

Adds a `benchmarks`-classifier jar to the blob and sts modules (provider +
client tiers) via a second `maven-jar-plugin` `test-jar` execution with an
include filter. The jar carries only:

- benchmark classes (`**/*Benchmark*.class`)
- JMH-generated classes (`**/jmh_generated/**`)
- JMH metadata (`META-INF/BenchmarkList`, `META-INF/CompilerHints`)

## Why

Our JMH benchmarks are owned and run by multicloudj — that doesn't change here. What changes is how they get executed.

A benchmark run can't happen inside the library's own build. It needs live cloud resources (real S3/GCS/DynamoDB, the actual substrate boundary, not mocks), a deployed environment so the numbers are stable and comparable, and a path into the metrics pipeline (Argus/Grafana, tracked per SDK version). That's a deploy-time workload, not a unit-test one, so it runs in a separate deployable benchmark-runner. Still ours — just not the library jar.

That runner is pinned to a released SDK version and bumped forward over time. The problem: today it carries the benchmark source hand-copied, so bumping the SDK version only swaps the library — the benchmark code actually being executed stays frozen at whatever was last copied in. Silent drift between the version under test and the benchmarks running against it.

Publishing the benchmark classes as a version-pinned artifact closes that gap: bumping the version swaps the real released benchmark code too, so the runner always executes the benchmarks that shipped with the SDK it's testing. No hand-copying, no sync to forget.

**Why a separate `benchmarks` jar and not the standard `:tests` jar?** Two reasons:

1. **Deployability** — in sfdc-bazel a `:tests` artifact is automatically treated as a `testonly` target, and a deployable target (like the benchmark runner) is not allowed to depend on a testonly target. A normal classifier jar has no such restriction.
2. **Security / CVEs** — the `:tests` jar pulls in the full test dependency set (WireMock, Mockito, Jetty, Netty), reintroducing CVEs the benchmark image deliberately strips out. Our filtered `benchmarks` jar carries only benchmark classes + JMH output, so its footprint stays clean (verified: zero of those libraries in any built jar).

## Scope

Two verticals (blob, sts) as a proof of the pattern. If the approach is
accepted, the same two-line-per-module change extends to the remaining
benchmark-bearing modules.

## Verification (local, JDK 25, `install -DskipTests`)

- `blob-aws-*-benchmarks.jar`: 45 classes, BenchmarkList = 66, zero forbidden deps
- `sts-aws-*-benchmarks.jar`: concrete benchmark + 5 JMH-generated classes, BenchmarkList = 10, zero forbidden deps, no `-tests.jar`
- `*-client-*-benchmarks.jar`: abstract bases only, empty list (base tier), clean deps
- A clean consumer (no vendored source) resolved main + benchmarks jars, JMH `-l` discovered all 66 blob benchmarks, and one method ran the full path to the external S3 boundary — identical to the vendored runner.

## Benchmark-jar guard (verify phase)

Following review, added a `maven-antrun-plugin` check (verify phase) to
blob-aws and sts-aws. It reads `META-INF/BenchmarkList` out of the built
`-benchmarks.jar` and fails the build if the list is missing/empty, or if it
contains no benchmark from the provider's package. A broken include filter
would otherwise pass every test yet publish a jar JMH discovers zero
benchmarks in. Runs on the `build-test-codecov` verify check and the release
deploy path (both `mvn verify`).

Two checks per module:
- BenchmarkList present and non-empty (else JMH finds zero benchmarks).
- BenchmarkList contains the provider package (`com.salesforce.multicloudj.blob.aws.`
  / `...sts.aws.`) — asserts the provider's benchmarks were actually packaged,
  and survives benchmark class renames.

Client modules are excluded by design: JMH only lists runnable benchmarks,
which live in the provider subclasses, so a `-client` benchmarks jar has a
legitimately empty BenchmarkList.

Guard testing:
- Positive (clean `install -DskipTests`, JDK 21): both jars build and the
  guard passes — blob-aws BenchmarkList = 66 lines, sts-aws = 10 lines, each
  containing its provider package.
- Negative (remove the benchmark include, `clean verify`): guard fails the
  build with "benchmarks jar has a missing or empty BenchmarkList - zero
  benchmarks".

## Notes

- Additive: does not alter the existing main or `:tests` jars, so it's independently revertible.
- Include glob `**/*Benchmark*.class` also catches meta/gating test classes whose names contain "Benchmark" (harmless — clean deps, not in BenchmarkList so JMH ignores them); can tighten during rollout.

### CODE REVIEW

File: blob/blob-aws/pom.xml [hunk 1]
```diff
@@ -707,6 +707,29 @@
...
+                        <id>benchmarks-classifier-jar</id>
+                        <goals>
+                            <goal>test-jar</goal>
+                        </goals>
+                        <configuration>
+                            <classifier>benchmarks</classifier>
+                            <includes>
+                                <include>**/*Benchmark*.class</include>
+                                <include>**/jmh_generated/**</include>
+                                <include>META-INF/BenchmarkList</include>
```
Comment (LihaoLiuXs) [inline blob/blob-aws/pom.xml:726]: Could we add an artifact-level CI check for this classifier? `BenchmarkList` is the discovery index JMH consumes, but the current tests never exercise the packaged jar. I removed only this `<include>` and ran the repository's `clean verify` path: the 629-test blob unit suite (2 skipped) and 133-test replay/integration suite (14 skipped) still completed successfully, while the classifier still contained the benchmark and generated classes but `org.openjdk.jmh.Main -l` discovered zero benchmarks. Please have CI inspect or run the packaged provider classifier—for example, assert a nonempty `META-INF/BenchmarkList` and that JMH lists an expected concrete provider benchmark—so an unusable artifact cannot be published successfully.

Comment (iamabhilaksh) [reply blob/blob-aws/pom.xml:726]: Good call — nothing was exercising the packaged jar, so a broken include filter would let all tests pass while JMH silently discovers zero benchmarks and we'd publish an unusable artifact.

Added a verify-phase maven-antrun guard to blob-aws and sts-aws. It reads META-INF/BenchmarkList from the built -benchmarks.jar and fails the build if the list is missing/empty, or if it contains no benchmark from the provider's package (com.salesforce.multicloudj.blob.aws. / ...sts.aws.). Asserting the package rather than a specific class keeps it robust to benchmark renames. It runs on the build-test-codecov verify check and the release deploy path, so a bad jar can't reach publish. Client modules are excluded — JMH only lists runnable benchmarks (provider subclasses), so a -client jar's list is legitimately empty.

Testing details are in the PR description under "Benchmark-jar guard (verify phase)" — validated both the positive pass and the negative failure.

Kept it inlined per-module for now to keep this a tight pattern proof; I'll centralize the guard in the follow-up PR that rolls the pattern out to all modules (with the expected package parameterized per provider).

**Comment by sandeepvinayak** [comment]
> Today the only way another service can run them against a given SDK version is to copy that test code by hand and keep it in sync — brittle and easy to forget. 

why does any other services need to run the benchmarks for multicloudj? Owning and executing these benchmarks are for multicloudj, can you please elaborate the intent?

**Comment by iamabhilaksh** [comment]
> > Today the only way another service can run them against a given SDK version is to copy that test code by hand and keep it in sync — brittle and easy to forget.
> 
> why does any other services need to run the benchmarks for multicloudj? Owning and executing these benchmarks are for multicloudj, can you please elaborate the intent?

Good catch @sandeepvinayak  — the phrase "another service" was misleading and made it sound like a downstream team consumes and runs our benchmarks. That's not the case. Let me reframe.

The benchmarks are owned and run by multicloudj — that doesn't change. What changes is how they get executed.

A benchmark run can't happen inside the library's own build. It needs live cloud resources (real S3/GCS/DynamoDB, the actual substrate boundary, not mocks), a deployed environment so the numbers are stable and comparable, and a path into the metrics pipeline (Argus/Grafana, tracked per SDK version). That's a deploy-time workload, not a unit-test one, so it runs in a separate deployable benchmark-runner. Still ours — just not the library jar.

That runner is pinned to a released SDK version and bumped forward over time. The problem: today it carries the benchmark source hand-copied, so bumping the SDK version only swaps the library — the benchmark code actually being executed stays frozen at whatever was last copied in. Silent drift between the version under test and the benchmarks running against it.

Publishing the benchmark classes as a version-pinned artifact closes that gap: bumping the version swaps the real released benchmark code too, so the runner always executes the benchmarks that shipped with the SDK it's testing. No hand-copying, no sync to forget.

**Review by LihaoLiuXs — APPROVED** [review]

File: blob/blob-aws/pom.xml [hunk 2]
```diff
@@ -707,6 +707,65 @@
...
+                                </loadresource>
+                                <fail message="benchmarks jar has a missing or empty BenchmarkList - zero benchmarks">
+                                    <condition>
+                                        <not><isset property="benchmarkList"/></not>
+                                    </condition>
+                                </fail>
+                                <fail message="benchmarks jar is missing com.salesforce.multicloudj.blob.aws benchmarks">
+                                    <condition>
+                                        <not>
+                                            <contains string="${benchmarkList}"
```
Comment (LihaoLiuXs) [inline blob/blob-aws/pom.xml:759]: This only validates provider `BenchmarkList` text, so the classifier bundle can still pass `verify` while being unrunnable. On this head I removed only the `**/*Benchmark*.class` include from `blob-client`; `mvn -o -B -pl blob/blob-aws -am clean verify -DskipTests -Djacoco.skip=true` stayed green and this check passed. The resulting client classifier had no `AbstractBlobBenchmarkTest`; the provider metadata still advertised all 66 benchmarks, but executing `AwsBlobBenchmarkTest.benchmarkGetMetadata` from the packaged provider+client pair failed with `NoClassDefFoundError: AbstractBlobBenchmarkTest`. The STS pair has the same dependency shape (`AwsStsBenchmarkTest` extends the unguarded client base). Could the verify guard exercise/class-link an expected benchmark using both published classifiers, or otherwise validate required classes in both jars, so this one-line include regression cannot publish an unusable bundle?

**Review by kchoy-sfdc — APPROVED** [review]
LGTM

**Comment by sandeepvinayak** [comment]
> > > Today the only way another service can run them against a given SDK version is to copy that test code by hand and keep it in sync — brittle and easy to forget.
> > 
> > 
> > why does any other services need to run the benchmarks for multicloudj? Owning and executing these benchmarks are for multicloudj, can you please elaborate the intent?
> 
> Good catch @sandeepvinayak — the phrase "another service" was misleading and made it sound like a downstream team consumes and runs our benchmarks. That's not the case. Let me reframe.
> 
> The benchmarks are owned and run by multicloudj — that doesn't change. What changes is how they get executed.
> 
> A benchmark run can't happen inside the library's own build. It needs live cloud resources (real S3/GCS/DynamoDB, the actual substrate boundary, not mocks), a deployed environment so the numbers are stable and comparable, and a path into the metrics pipeline (Argus/Grafana, tracked per SDK version). That's a deploy-time workload, not a unit-test one, so it runs in a separate deployable benchmark-runner. Still ours — just not the library jar.
> 
> That runner is pinned to a released SDK version and bumped forward over time. The problem: today it carries the benchmark source hand-copied, so bumping the SDK version only swaps the library — the benchmark code actually being executed stays frozen at whatever was last copied in. Silent drift between the version under test and the benchmarks running against it.
> 
> Publishing the benchmark classes as a version-pinned artifact closes that gap: bumping the version swaps the real released benchmark code too, so the runner always executes the benchmarks that shipped with the SDK it's testing. No hand-copying, no sync to forget.

sounds good, just a minor thing that we need to refer bazel stuff in here.

**Review by sandeepvinayak — APPROVED** [review]

**Comment by iamabhilaksh** [comment]
Thanks @sandeepvinayak. The runner-side consumption — how the pinned classifier actually gets resolved and bumped — is the follow-up that's tracked separately from this PR; I'll make sure that mechanism is spelled out there. Keeping this one scoped to just publishing the artifact so the two concerns stay cleanly split.

File: blob/blob-aws/pom.xml (hunk 2 above)
Comment (iamabhilaksh) [reply blob/blob-aws/pom.xml:759]: Nice repro — thanks for pushing on this. You're right: the guard only asserts the provider's `BenchmarkList` text, so it proves JMH would *discover* the benchmarks, not that they'd actually link and run. A missing base class in the client classifier sails straight through.

One thing that shifts in the centralized follow-up (#606): the include filter you edited on `blob-client` no longer lives per-module — it's defined once in the root `pluginManagement`, so there's no per-client include left to drop. But that doesn't close your real point — a client that doesn't opt in, or a base-class rename, still produces a provider bundle whose superclass isn't published anywhere while `verify` stays green.

So I'll add the linkability check you're describing to the centralized guard: at the provider's `verify`, put both the provider and client `-benchmarks` classifiers on a classpath and force-load an expected benchmark through its full supertype chain, failing the build on `NoClassDefFoundError` — exercising it the way you did by hand, so it can't publish an unrunnable bundle regardless of which include or opt-in slipped. Let me validate the approach in the reactor and I'll fold it into #606 where the guard now lives; will follow up here once it's in.

Comment (iamabhilaksh) [reply blob/blob-aws/pom.xml:759]: Following up — and correcting course from what I said above.

I built the class-link check and validated it in the reactor, and the "force-load a benchmark through its supertype chain" approach turned out heavier and more fragile than I made it sound. `Class.forName(init=true)` links the *entire* transitive graph — the moment you load a benchmark it wants SLF4J, JMH, the AWS SDK, all of it. On a classpath of just the two `-benchmarks` classifiers it dies with `NoClassDefFoundError: org/slf4j/LoggerFactory`, not on the missing base class. To link cleanly you'd have to assemble each provider's full runtime classpath while excluding first-party build outputs (so you test the *published* jars, not the build classpath), for every provider. That's a lot of moving parts in a build guard, and one that can false-fail on a missing transitive dep is worse than the gap it closes.

So I went with a check that gives the same guarantee without the classpath fragility: at `verify`, every benchmark-pattern class the module *compiled* (`**/*Benchmark*.class`, `**/jmh_generated/**`) must be present in its published `-benchmarks.jar`. The jar is built from the compiled test-classes through the include filter, so it's always a subset — equality is the correctness condition, and a dropped include makes compiled-count exceed jar-count and fails the build. Your exact repro now trips it:

`sts-client: benchmarks jar has 0 benchmark classes but 5 were compiled - the include filter dropped classes; refusing to publish an incomplete bundle`

Validated the full matrix on a clean reactor build: all 21 benchmark jars pass (compiled == jar), the dropped-include case fails as above, and non-benchmark modules / `-Dmaven.test.skip` no-op.

On the residuals I raised earlier — in #606 the include lives once in the root `pluginManagement` and every client opts in, so "client didn't opt in" is structurally closed for the current set, and this check guarantees no compiled base class can be silently dropped from any bundle. What it *doesn't* catch is a future provider whose client forgets to opt in entirely (no jar built, nothing to compare). If we want the guard to enforce cross-module supertype reachability there, the clean way is a `javap`-based check that reads each benchmark's `extends` from the constant pool and asserts any first-party superclass is published in one of the classifiers — first-party only, no runtime classpath. Happy to add that as a follow-up if you think it's worth it; for the regression you demonstrated, this closes it.

It's in #606, where the centralized guard now lives.
